src/server/api/osdi/api-auth.js

```diff
@@ -0,0 +1,62 @@
+
+import { r } from '../../models';
+import { getHash } from './guid'
+
+function sendUnauthorizedResponse(res) {
+  res.writeHead(401, { 'WWW-Authenticate': 'Basic realm=Contacts API' })
+  res.end('Unauthorized')
+}
+
+function publicApi(features) {
+  if (features.publicApi === true || process.env.PUBLIC_API === 'true') {
+    return true
+  } else {
+    return false
+  }
+}
+
+export async function authShortCircuit(req, res, orgId) {
+
+  const organization = await r.knex('organization').where('id', orgId).first('features')
+  const features = organization.features ? JSON.parse(organization.features) : {}
+  const apiKey = features.apiKey;
+
+  if (publicApi(features) === true) {
+    return false
+  }
+
+  if (!(('authorization' in req.headers)
+      || ('osdi-api-token' in req.headers))) {
+    sendUnauthorizedResponse(res)
```

Comment:
Can we also bail early if apiKey isn't present?

Also, I worry a bit about the difference in response if we enable this feature or not -- maybe we should be returning 404 in these cases?  Thoughts?

Reply:
this changed too.  for an auto failure we just return 403 Forbidden.  If OSDI is disabled, it will return 510.  Both include a json body with hints on why.  eg:
__Access Denied - 403 Forbidden__
```
{
  "osdi:error": {
    "response_code": 403,
    "error_description": "You need to use an OSDI-API-Token HTTP header containing your token, or you need to be logged in with sufficient privileges.",
    "auth_status": {
      "token_status": "INVALID_OSDI_TOKEN",
      "user_status": "LOGGED_OUT",
      "authenticated": false,
      "osdi_enabled": true
    }
  }
}
```
__OSDI Disabled - 510 Not Extended__
```
{
  "osdi:error": {
    "response_code": 510,
    "error_description": "OSDI is not enabled for this resource"
  }
}
```